utils: Add new design for generator script

razvand opened the pull request:
Add new design for generator script

razvand committed:
utils: Add new design for generator script

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

RaduNichita commented on utils/new-design/all.py:
```diff
@@ -0,0 +1,194 @@
+import itertools
+
+class ConfigVariants:
+    arch = ['arm64', 'x86_64']
```
```suggestion
    arch = ['arm64', 'x86_64']
```

I think it would be better to have a dictionary here to keep track a number of options (an option means arch, hypervisor, platform etc)

RaduNichita commented on utils/new-design/all.py:
```diff
@@ -0,0 +1,194 @@
+    networking = ['none', 'bridge', 'nat', 'tap']
+
+    def _convert_to_dict(i):
+        return {
```
```suggestion
        return {
```

could you add an assert for the number of options, to make sure we get an error when we add a new option

razvand committed:
Document new class design

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Improve tester config

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils: Generate target configs for new design

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils/new-design: Add initial generator logic

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add Kraftfile generator

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add Makefile generator

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add generators for build scripts

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Complete build scripts

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Separate build variants from run variants

Until this point a variant consisted of a combination of build and run
parameters. However, this was redundant as the the build parameters (and
build) were part of multiple variants.

Separate build parameters from run parameters in `tester.yaml`. A
variant consists of a build variant (set of parameters) and multiple run
variants (an array of sets of parameters).

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add support for Clang

Update templates for build commands to use Clang or GCC.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add support for building Xen

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-desing: Add run scripts

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add support for ports for NAT networking

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Consistent naming for networking types

Use the same name networking in the tester configuration file and in the
template script names: "bridge", "nat", "tap".

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add Firecracker run support

Add Firecracker template scripts and support in generating run scripts
and configuration files (JSON).

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add common directory

Add directory with common support scripts. Add `common.sh` script used
for running tests.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Use machine variable for non-arch runs

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
library/nginx/1.15: Add setup for new design scripts

Add `xen` to Kraftfile.
Add `config.yaml` with testing configurations.
Add `.test` basic test script.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
library/nginx/1.15: Add test scripts

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
library/nginx/1.15: Ignore .tests/ directory

`.tests/` is the working directory for testing.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
native/http-c: Add testing scripts

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils/new-design: Remove run configs when rootfs is not present

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils/new-design: Pass configuration file as argument

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils/new-design: Fixes for when rootfs is absent

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils/new-design: Fixes for base build

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils: Update tst scripts

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils: Use example targets

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
new-design: Add fs build (init) script

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils: Add comments to functions in all.py script

Add documentation as comments to all.py script that generates target
configuration files and scripts.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils: Rename *Config to *Setup classes

Rename classes from TargetConfig, BuildConfig and RunConfig to
TargetSetup, BuildSetup and RunSetup. These are more appropriate names,
since the classes are used to both store configuration information and
generate required configuration files and scripts.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
utils: Refactor Python script into multiple scripts

Refactor `all.py` script into multiple scripts. Each script stores its
own class: `TesterConfig`, `SystemConfig`, `AppConfig`, `TargetSetup`,
`BuildSetup`, `RunSetup`.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>
